src/Platform/Microsoft.Testing.Platform/DotnetTest/DotnetTestConnection.cs

```diff
@@ -0,0 +1,167 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Globalization;
+using System.Runtime.InteropServices;
+
+using Microsoft.Testing.Platform.CommandLine;
+using Microsoft.Testing.Platform.Extensions.CommandLine;
+using Microsoft.Testing.Platform.Helpers;
+using Microsoft.Testing.Platform.IPC;
+using Microsoft.Testing.Platform.IPC.Models;
+using Microsoft.Testing.Platform.IPC.Serializers;
+using Microsoft.Testing.Platform.Services;
+using Microsoft.Testing.Platform.Tools;
+
+namespace Microsoft.Testing.Platform;
+
+internal sealed class DotnetTestConnection :
+#if NETCOREAPP
+    IAsyncDisposable,
+#endif
+    IDisposable
+{
+    private readonly IServiceProvider _serviceProvider;
+    private readonly ITestApplicationCancellationTokenSource _cancellationTokenSource;
+
+    private NamedPipeClient? _dotnetTestPipeClient;
+
+    private CommandLineHandler CommandLineHandler => _serviceProvider.GetRequiredService<CommandLineHandler>();
+
+    private IEnvironment Environment => _serviceProvider.GetRequiredService<IEnvironment>();
+
+    private ITestApplicationModuleInfo TestApplicationModuleInfo => _serviceProvider.GetRequiredService<ITestApplicationModuleInfo>();
+
+    public DotnetTestConnection(IServiceProvider serviceProvider, ITestApplicationCancellationTokenSource cancellationTokenSource)
+    {
+        _serviceProvider = serviceProvider;
+        _cancellationTokenSource = cancellationTokenSource;
+    }
+
+    public async Task<bool> TryConnectToDotnetTestPipeIfAvailableAsync()
+    {
+        // If we are in server mode and the pipe name is provided
+        // then, we need to connect to the pipe server.
+        if (CommandLineHandler.HasDotnetTestServerOption() &&
+            CommandLineHandler.TryGetOptionArgumentList(PlatformCommandLineProvider.DotNetTestPipeOptionKey, out string[]? arguments))
+        {
+            // The execution id is used to identify the test execution
+            // We are storing it as an env var so that it can be read by the test host, test host controller and the test host orchestrator
+            // If it already exists, we don't overwrite it
+            if (RoslynString.IsNullOrEmpty(Environment.GetEnvironmentVariable(EnvironmentVariableConstants.TESTINGPLATFORM_DOTNETTEST_EXECUTIONID)))
+            {
+                Environment.SetEnvironmentVariable(EnvironmentVariableConstants.TESTINGPLATFORM_DOTNETTEST_EXECUTIONID, Guid.NewGuid().ToString("N"));
+            }
+
+            _dotnetTestPipeClient = new(arguments[0]);
+            _dotnetTestPipeClient.RegisterAllSerializers();
+
+            await _dotnetTestPipeClient.ConnectAsync(_cancellationTokenSource.CancellationToken);
+
+            return _dotnetTestPipeClient.IsConnected;
+        }
+
+        return false;
+    }
+
+    public async Task SendCommandLineOptionsToDotnetTestPipeAsync()
+    {
+        if (_dotnetTestPipeClient?.IsConnected == false)
```

Comment:
it's not expected this call or? I prefer an opposite RoslynAssert  `_dotnetTestPipeClient?.IsConnected == true`